src/axom/primal/geometry/BezierCurve.hpp

```diff
@@ -0,0 +1,335 @@
+// Copyright (c) 2017-2019, Lawrence Livermore National Security, LLC and
+// other Axom Project Developers. See the top-level COPYRIGHT file for details.
+//
+// SPDX-License-Identifier: (BSD-3-Clause)
+
+/*!
+ * \file BezierCurve.hpp
+ *
+ * \brief A BezierCurve primitive
+ */
+
+#ifndef PRIMAL_BEZIERCURVE_HPP_
+#define PRIMAL_BEZIERCURVE_HPP_
+
+#include "axom/core.hpp"
+#include "axom/slic.hpp"
+
+#include "axom/primal/geometry/NumericArray.hpp"
+#include "axom/primal/geometry/Point.hpp"
+#include "axom/primal/geometry/Vector.hpp"
+#include "axom/primal/geometry/Segment.hpp"
+#include "axom/primal/geometry/BoundingBox.hpp"
+#include "axom/primal/geometry/OrientedBoundingBox.hpp"
+
+#include "axom/primal/operators/squared_distance.hpp"
+
+#include "fmt/fmt.hpp"
+#include <vector>
+#include <ostream>
+
+namespace axom
+{
+namespace primal
+{
+
+// Forward declare the templated classes and operator functions
+template < typename T, int NDIMS >
+class BezierCurve;
+
+/*! \brief Overloaded output operator for Bezier Curves*/
+template < typename T,int NDIMS >
+std::ostream& operator<<(std::ostream & os,
+                         const BezierCurve< T,NDIMS > & bCurve);
+
+/*!
+ * \class BezierCurve
+ *
+ * \brief Represents a Bezier curve defined by an array of control points
+ * \tparam T the coordinate type, e.g., double, float, etc.
+ * \tparam NDIMS the number of dimensions
+ *
+ * The order of a Bezier curve with N+1 control points is N.
+ * The curve is approximated by the control points,
+ * parametrized from t=0 to t=1.
+ */
+
+template < typename T,int NDIMS >
+class BezierCurve
+{
+public:
+  using PointType = Point< T,NDIMS >;
+  using VectorType = Vector< T,NDIMS >;
+  using NumArrayType = NumericArray< T,NDIMS >;
+  using SegmentType = Segment< T, NDIMS >;
+  using CoordsVec = std::vector< PointType >;
```

Comment:
I understand the convenience of STL, but, we should avoid it.

Could we just use `axom::Array`, or `numerics::Matrix` with dimensions `ndims x numPoints`, i.e., each column in the matrix is a control point (node), or just flat arrays?

Reply:
Good suggestion @gzagaris.

Let's save stl changes for a future PR since @davidgunderman has another branch with further changes.